File: src/main/java/com/thebluehats/server/game/managers/enchants/Timer.java
package com.thebluehats.server.game.managers.enchants;

import com.google.inject.Inject;
import org.bukkit.Bukkit;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.HashMap;
import java.util.Set;

public class Timer<K> {
    private final HashMap<K, TimerData> timerData = new HashMap<>();

    private final JavaPlugin plugin;

    @Inject
    public Timer(JavaPlugin plugin) {
        this.plugin = plugin;
    }

    public void start(K key, long ticks) {
        TimerData data = timerData.computeIfAbsent(key, k -> new TimerData());

        data.setTime(ticks);

        data.setTaskId(Bukkit.getServer().getScheduler().scheduleSyncRepeatingTask(plugin, () -> {
            data.setTime(data.getTime() - 1);

            if (data.getTime() <= 0) {
                data.setTime(0);

                Bukkit.getServer().getScheduler().cancelTask(data.getTaskId());
                timerData.remove(key);
            }
        }, 0L, 1L));
    }

    public void start(K key, long ticks, Runnable post) {
        TimerData data = timerData.computeIfAbsent(key, k -> new TimerData());

        data.setTime(ticks);

        data.setTaskId(Bukkit.getServer().getScheduler().scheduleSyncRepeatingTask(plugin, () -> {
            data.setTime(data.getTime() - 1);

            if (data.getTime() <= 0) {
                data.setTime(0);

                Bukkit.getServer().getScheduler().cancelTask(data.getTaskId());
                timerData.remove(key);

                post.run();
            }
        }, 0L, 1L));
    }

    public void cancel(K key) {
        TimerData data = timerData.computeIfAbsent(key, k -> new TimerData());

        Bukkit.getServer().getScheduler().cancelTask(data.getTaskId());
        timerData.remove(key);
    }

    public Set<K> getKeys() {
        return timerData.keySet();
    }

    public boolean isRunning(K key) {
        TimerData data = timerData.get(key);

        return data != null && data.isRunning();
    }

    public long getTime(K key) {
        TimerData data = timerData.get(key);

        return data != null ? data.getTime() : 0L;
    }

    private static class TimerData {
        private long cooldownTime;
        private int cooldownTaskId;

        public boolean isRunning() {
            return getTime() != 0;
        }

        public long getTime() {
            return cooldownTime;
        }

        public void setTime(long value) {
            cooldownTime = value;
        }

        public int getTaskId() {
            return cooldownTaskId;
        }

        public void setTaskId(int value) {
            cooldownTaskId = value;
        }
    }
}

File: src/main/kotlin/com/thebluehats/server/game/managers/enchants/Timer.kt
package com.thebluehats.server.game.managers.enchants

import com.google.inject.Inject
import org.bukkit.Bukkit
import org.bukkit.plugin.java.JavaPlugin
import java.util.*

class Timer<K> @Inject constructor(private val plugin: JavaPlugin) {
    private val timerData = HashMap<K, TimerData>()

    fun start(key: K, time: Long, resetTime: Boolean = false, seconds: Boolean = false, post: Runnable? = null) {
        val data = timerData.computeIfAbsent(key, { TimerData() })

        if (data.isRunning) {
            if (resetTime) data.time = if (seconds) time * 20 else time

            return
        }

        data.time = time

        data.taskId = Bukkit.getServer().scheduler.scheduleSyncRepeatingTask(plugin, {
            data.time = data.time - 1

            if (data.time <= 0) {
                data.time = 0
                Bukkit.getServer().scheduler.cancelTask(data.taskId)
                timerData.remove(key)

                post?.run()
            }
        }, 0L, 1L)
    }

    fun cancel(key: K) {
        val data = timerData.computeIfAbsent(key, { TimerData() })
        Bukkit.getServer().scheduler.cancelTask(data.taskId)
        timerData.remove(key)
    }

    val keys: Set<K>
        get() = timerData.keys

    fun isRunning(key: K): Boolean {
        val data = timerData[key]
        return data != null && data.isRunning
    }

    fun getTime(key: K): Long {
        val data = timerData[key]
        return data?.time ?: 0L
    }

    private class TimerData {
        var time: Long = 0
        var taskId = 0
        val isRunning: Boolean
            get() = time != 0L
    }
}